basesandbox/src/ipc/domain_socket.rs

```diff
@@ -0,0 +1,166 @@
+        std::fs::remove_file(&self.address_server).unwrap();
+        std::fs::remove_file(&self.address_client).unwrap();
+    }
+}
+
+// TODO: Use stream instead of datagram.
+struct SocketInternal(UnixDatagram);
+impl Drop for SocketInternal {
+    fn drop(&mut self) {
+        self.0.shutdown(std::net::Shutdown::Both).unwrap();
+    }
+}
+
+pub struct DomainSocketSend {
+    address_dst: String,
+    socket: Arc<SocketInternal>,
+}
+
+impl IpcSend for DomainSocketSend {
+    fn send(&self, data: &[u8]) {
+        assert_eq!(self.socket.0.send_to(data, &self.address_dst).unwrap(), data.len());
+    }
+}
+
+pub struct DomainSocketRecv {
+    address_dst: String,
+    socket: Arc<SocketInternal>,
+    buffer: RefCell<Vec<u8>>,
+}
+
+impl IpcRecv for DomainSocketRecv {
+    type MyTerminate = Terminator;
+
+    fn recv(&self, timeout: Option<std::time::Duration>) -> Result<Vec<u8>, RecvFlag> {
+        self.socket.0.set_read_timeout(timeout).unwrap();
+        let (count, address) = self.socket.0.recv_from(&mut self.buffer.borrow_mut()).unwrap();
+        assert!(count <= self.buffer.borrow().len(), "Unix datagram got data larger than the buffer.");
+        if count == 0 {
+            return Err(RecvFlag::Termination)
+        }
+        assert_eq!(
+            address.as_pathname().unwrap(),
+            Path::new(&self.address_dst),
+            "Unix datagram received packet from an unexpected sender."
+        );
+        Ok(self.buffer.borrow()[0..count].to_vec())
+    }
+
+    fn create_terminate(&self) -> Self::MyTerminate {
+        Terminator(self.socket.clone())
+    }
+}
+
+pub struct Terminator(Arc<SocketInternal>);
+
+impl Terminate for Terminator {
+    fn terminate(&self) {
+        (self.0).0.shutdown(std::net::Shutdown::Both).unwrap();
+    }
+}
+
+pub struct DomainSocket {
+    send: DomainSocketSend,
+    recv: DomainSocketRecv,
+}
+
+impl InterProcessUnit for DomainSocket {
+    fn new(data: Vec<u8>) -> Self {
+        let (address_src, address_dst): (String, String) = serde_cbor::from_slice(&data).unwrap();
+        let socket = Arc::new(SocketInternal(UnixDatagram::bind(&address_src).unwrap()));
+        DomainSocket {
+            send: DomainSocketSend {
+                address_dst: address_dst.clone(),
+                socket: socket.clone(),
+            },
+            recv: DomainSocketRecv {
+                address_dst,
+                socket,
+                buffer: RefCell::new(vec![0; 1024]),
+            },
+        }
+    }
+}
+
+impl IpcSend for DomainSocket {
+    fn send(&self, data: &[u8]) {
+        self.send.send(data)
+    }
+}
+
+impl IpcRecv for DomainSocket {
+    type MyTerminate = Terminator;
+
+    fn recv(&self, timeout: Option<std::time::Duration>) -> Result<Vec<u8>, RecvFlag> {
+        self.recv.recv(timeout)
+    }
+
+    fn create_terminate(&self) -> Self::MyTerminate {
+        self.recv.create_terminate()
+    }
+}
+
+impl Ipc for DomainSocket {
+    type SendOnly = DomainSocketSend;
```

Comment:
I would have named them as `SenderHalf` and `ReceiverHalf`.

Reply:
That's better. Thanks